src/Cabal2nix.hs

```diff
@@ -0,0 +1,176 @@
+          <*> optional (strOption $ long "subpath" <> metavar "PATH" <> help "Path to Cabal file's directory relative to the URI (default is root directory)")
+          <*> strArgument (metavar "URI")
+
+readP :: P.ReadP a a -> ReadM a
+readP p = eitherReader $ \s -> case [ r' | (r',"") <- P.readP_to_S p s ] of
+                                    (r:_) -> Right r
+                                    _     -> Left ("invalid value " ++ show s)
+
+parsePlatform :: P.ReadP r Platform
+parsePlatform = do arch <- P.choice [P.string "i686" >> return I386, P.string "x86_64" >> return X86_64]
+                   _ <- P.char '-'
+                   os <- P.choice [P.string "linux" >> return Linux, P.string "darwin" >> return OSX]
+                   return (Platform arch os)
+
+pinfo :: ParserInfo Options
+pinfo = info
+        (   helper
+        <*> infoOption ("cabal2nix " ++ display version) (long "version" <> help "Show version number")
+        <*> options
+        )
+        (  fullDesc
+        <> header "cabal2nix converts Cabal files into build instructions for Nix."
+        <> progDescDoc (Just (P2.vcat
+                     [ P2.text ""
+                     , P2.text "Recognized URI schemes:"
+                     , P2.text ""
+                     , P2.text "  cabal://pkgname-pkgversion     download the specified package from Hackage"
+                     , P2.text "  cabal://pkgname                download latest version of this package from Hackage"
+                     , P2.text "  file:///local/path             load the Cabal file from the local disk"
+                     , P2.text "  /local/path                    abbreviated version of file URI"
+                     , P2.text "  <git/svn/bzr/hg URL>           download the source from the specified repository"
+                     , P2.text ""
+                     , P2.fillSep (map P2.text (words (  "If the URI refers to a cabal file, information for building the package "
+                                                      ++ "will be retrieved from that file, but hackage will be used as a source "
+                                                      ++ "for the derivation. Otherwise, the supplied URI will be used to as the "
+                                                      ++ "souce for the derivation and the information is taken from the cabal file "
+                                                      ++ "at the root of the downloaded source."
+                                                      )))
+                     ]))
+        )
+
+
+safeFlush :: IO () -> IO ()
+safeFlush io = bracket (return ()) (\() -> hFlush stdout >> hFlush stderr) $ \() -> io
+
+main :: IO ()
+main = mainWithArgs =<< getArgs
+
+mainWithArgs :: [String] -> IO ()
+mainWithArgs args = safeFlush $ do
+  Options {..} <- handleParseResult $ execParserPure defaultPrefs pinfo args
```

Comment:
If we'd like to do regression testing of `cabal2nix` via the command-line interface, then having such a re-usable module makes sense. BTW, if you ever want to call a `main` function with specific arguments from a Haskell library, then `System.Environment.withArgs` can do it.